### PR TITLE
UBER-2664 Add Plugin Logger documentation.

### PR DESCRIPTION
**Pull request by donoage**
* Added Plugin Logger documentation.
* Added **table of contents** for big markdown files.
* Click on `view file` button for each file to review.

### CODE REVIEW

**Review by michaelpwest — APPROVED** [review]

File: docs/DEVELOPMENT.md [hunk 1]
```diff
@@ -4,18 +4,41 @@ A plugin is a collection of files, functions, hooks, SDK methods and calls, and
...
+        - [Text search](#text-search)
+        - [Clearing the log view](#clearing-the-log-view)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Requirements
 A plugin requires a `manifest.json` and `bootstrap.php` file.
 
-### Optional
-* Additional PHP files that the bootstrap file includes.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md:11]: no period at the end

File: docs/DEVELOPMENT.md [hunk 2]
```diff
@@ -4,18 +4,41 @@ A plugin is a collection of files, functions, hooks, SDK methods and calls, and
...
-* Datasource classes.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md:12]: no period at the end

File: docs/DEVELOPMENT.md [hunk 3]
```diff
@@ -4,18 +4,41 @@ A plugin is a collection of files, functions, hooks, SDK methods and calls, and
...
-* JavaScript files.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md:13]: no period at the end

File: docs/DEVELOPMENT.md [hunk 4]
```diff
@@ -4,18 +4,41 @@ A plugin is a collection of files, functions, hooks, SDK methods and calls, and
...
-* CSS files.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md:14]: no period at the end

File: docs/DEVELOPMENT.md [hunk 5]
```diff
@@ -4,18 +4,41 @@ A plugin is a collection of files, functions, hooks, SDK methods and calls, and
...
-* Any additional files the plugin needs (for example, images).
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md:15]: no period at the end

File: docs/DEVELOPMENT.md [hunk 6]
```diff
@@ -58,15 +81,24 @@ Below is our example plugin manifest file:
...
 The bootstrap file, `bootstrap.php`, is a PHP file the plugin executes first. This file contains functions that implement defined hooks throughout Ubersmith. The bootstrap can also include other PHP files, to divide the code into multiple PHP files.  For more info, [see Bootstrap]().
 
 The bootstrap file requires a namespace:
 ```php
 namespace Docs\ClientTickets;
 ```
 
-## SDK Functions
+### Optional files
+* Additional PHP files that the bootstrap file includes.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: no period at the end

File: docs/DEVELOPMENT.md [hunk 7]
```diff
@@ -58,15 +81,24 @@ Below is our example plugin manifest file:
...
+* Datasource class files.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: no period at the end

File: docs/DEVELOPMENT.md [hunk 8]
```diff
@@ -58,15 +81,24 @@ Below is our example plugin manifest file:
...
+* JavaScript files.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: no period at the end

File: docs/DEVELOPMENT.md [hunk 9]
```diff
@@ -58,15 +81,24 @@ Below is our example plugin manifest file:
...
+* CSS files.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: no period at the end

File: docs/DEVELOPMENT.md [hunk 10]
```diff
@@ -58,15 +81,24 @@ Below is our example plugin manifest file:
...
+* Any additional files the plugin needs (for example, images).
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: no period at the end

File: docs/DEVELOPMENT.md [hunk 11]
```diff
@@ -118,10 +150,12 @@ if (empty($tickets)) {
 }
 ```
 
-## Hooks
+## Writing your plugin
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: Capitalize  Your and Plugin

File: docs/DEVELOPMENT.md [hunk 12]
```diff
@@ -184,10 +218,10 @@ $output .= '
 return $output;
 ```
 
-## Datasource
+### Datasource
 A datasource is a specialized plugin module whose purpose is to collect usage metrics for defined resources.  Any entity or device can be considered a resource of a datasource.  The definition and it's manner of data retrieval is left entirely up to the author of the plugin.  For example, a datasource can define switch ports as a resource to collect bandwidth bits -OR- an SMTP server as a resource to tally outbound emails.  Usage metrics collected are used by an assigned Service Plan allowing relevant Services to calculate a billable cost.
 
-A datasource requires a class implementing the `UbersmithSDK\Usage\Data\Source` interface as well as some annotations. For more information about the functions available to use, [see Datasources](). 
+A datasource requires a class implementing the `UbersmithSDK\Usage\Data\Source` interface as well as some annotations. For more information about the functions available to use, [see Datasources]().
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: A datasource requires a class implementing the `UbersmithSDK\Usage\Data\Source` interface as well as some annotations. For more information about the available functions, [see Datasources]().

File: docs/DEVELOPMENT.md [hunk 13]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
 require_once 'class.usage_datasource.php';
 ```
 
-### <a name="abcd"></a> Resource types
+### Resource types
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: Capitalize Types

File: docs/DEVELOPMENT.md [hunk 14]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
 The markup resource requires only a name and an identifier.
 This type of resource is used when the external system already calculates the price of usage. It will allow
 the user to select a markup, either as a percentage or a fixed amount, to be added to the price calculated by the external system.
 
 When using this resource type, the `fetch` method will return prices directly.
 
+## Development Tools
+
+### Plugin Logger
+Plugin Logger is available for use to track what PluginSystem hooks are being fired and how your plugins are reacting to them.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: Plugin Logger tracks what PluginSystem hooks are being fired and how your plugins are reacting to them.

File: docs/DEVELOPMENT.md [hunk 15]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+- Currently, the number of log entries are limited to the latest `10,000` events.
+- At least one plugin must be installed and set up with a module instance configuration.
+
+#### Setup
+- In your `config.ini.php` file under `[ubersmith]` block. Add the following line.
+    ```
+    plugin_logger_redis_path="tcp://redis:6379?database=3"
+    ```
+- In your Ubersmith application, go to _Settings_, _Plugins_, _Settings_ and enable Plugin Logger. Go back to _Plugins_ page.
+- Now the logger is enabled and you should see _Go to Logger_ link to access the logger.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: - Now the logger is enabled and you should see the _Go to Logger_ link to access the logger.

File: docs/DEVELOPMENT.md [hunk 16]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+
+#### Log Entry Definitions
+| Key | Value |
+| --- | --- |
+| `datetime` | datetime of log event. |
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: no period at end

File: docs/DEVELOPMENT.md [hunk 17]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+| `message` | message is composed of `[event_origin] event_description` |
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: need period at end.

File: docs/DEVELOPMENT.md [hunk 18]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+| `uid` | uid represents a single execution flow. For example, if you do a text search with a uid, the logger will display a group of events which transpired in the single execution. |
+| `event` | `hook_type` and `hook_name` related to the event. |
+| `plugin` | `identifier` from specified in manifest.json, the plugin's `module_name` and `function` related to the event. |
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: seems a word is missing here... what is specified?

File: docs/DEVELOPMENT.md [hunk 19]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+| `hostname` | For Ubersmith application with multiple web heads, this can act as a supplement to `uid`. |
+
+#### Log Entry Types and Colors
+- A log message with ![#008000](https://placehold.it/15/008000/000000?text=+) `GREEN` background color indicates a successful execution of a plugin function.
+- A log message with ![#f03c15](https://placehold.it/15/f03c15/000000?text=+) `RED` background color indicates a successful execution of a plugin function.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: does both green and red mean "success"? Is read failed?

File: docs/DEVELOPMENT.md [hunk 20]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+
++   Debug\Log('Simple string', 'a simple log');
++   Debug\Log('Log object', $config);
++   Debug\Log('Log anything', [$config, $tickets, 9999999999, 'qweqwe']);
+
+    ...
+}
+```
+
+###### View options
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: Capitalize Options

File: docs/DEVELOPMENT.md [hunk 21]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+- Toggle collapsed/expanded view modes with `view collapsed/expanded` button.
+
+###### Filters
+- You can filter per plugin and/or per usage on left sidebar navigation radio buttons.
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: change on to with

File: docs/DEVELOPMENT.md [hunk 22]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+
+###### Text search
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: Capitalize Search

File: docs/DEVELOPMENT.md [hunk 23]
```diff
@@ -265,25 +299,116 @@ Finally, the datasource class will need to be included in the bootstrap file:
...
+- Any text in any log entry is searchable.
+
+###### Clearing the log view
```
Comment (camyangelo) [inline docs/DEVELOPMENT.md]: Capitalize Log and View

File: docs/functions/FUNCTIONS.md [hunk 24]
```diff
@@ -752,4 +817,19 @@ URL encode a string.
 
 **Return value:** `string URL encoded string.`
 
+## Debug functions
```
Comment (camyangelo) [inline docs/functions/FUNCTIONS.md]: Capitalize Functions

**Review by camyangelo — CHANGES_REQUESTED** [review]
Have inline comments for you.

**Comment by donoage** [comment]
@camyangelo All updates have been made. Thanks Camille 🎉  Let me know if there's anything else.

**Comment by camyangelo** [comment]
@donoage - you are most welcome!

**Review by camyangelo — APPROVED** [review]

**Pull request closed without merging**